### PR TITLE
Add a function to qchem for two-mode squeezed state sampling

### PR DESCRIPTION
**Pull request by AroosaIjaz**
This PR adds the function `sample_tmsv` to `sf.apps.qchem.dynamics` to sample from a two-mode squeezed input state ( see ADR 101)

### CODE REVIEW

**Comment by AroosaIjaz** [comment]
Sorry. For some reason, some of the older commits also got in. They do not hurt but I wish they did not create this clutter here :D

File: strawberryfields/apps/qchem/dynamics.py [hunk 1]
```diff
@@ -250,3 +254,81 @@ def prob(samples: list, excited_state: list) -> float:
...
+
+        op(t, Ul, w) | q[:N]
+
+        if loss:
+            for _q in q:
+                sf.ops.LossChannel(1 - loss) | _q
+
+        sf.ops.MeasureFock() | q
+
+    with warnings.catch_warnings():
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:329]: What is the purpose of this?

File: strawberryfields/apps/qchem/dynamics.py [hunk 2]
```diff
@@ -250,3 +254,81 @@ def prob(samples: list, excited_state: list) -> float:
...
+        warnings.filterwarnings("ignore", category=UserWarning, message="Cannot simulate non-")
+
+        s = eng.run(prog, shots=n_samples).samples
+
+    return s.tolist()
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:334]: This comment is true also for PR #423 : Why do we convert to list here?

Comment (AroosaIjaz) [reply strawberryfields/apps/qchem/dynamics.py:334]: list of samples is a more elegant output. `eng.run(prog).samples` returns an array of samples with Gaussian backend. It returns even more things when fock backend is used. so for PR #423 , we have to further process the engine result. The user only sees the list of samples that are really easy to use further.

File: strawberryfields/apps/qchem/dynamics.py (hunk 1 above)
Comment (AroosaIjaz) [reply strawberryfields/apps/qchem/dynamics.py:329]: yeah I am not sure why this happens but we get a warning with the output when we run:
`xanadu/strawberryfields/strawberryfields/backends/gaussianbackend/backend.py:225: UserWarning: Cannot simulate non-Gaussian states. Conditional state after Fock measurement has not been updated.`
So Soran suppressed the warning. I do not know why it comes though.

File: strawberryfields/apps/qchem/dynamics.py [hunk 3]
```diff
@@ -250,3 +254,84 @@ def prob(samples: list, excited_state: list) -> float:
         raise ValueError("The excited state must not contain negative values")
 
     return samples.count(excited_state) / len(samples)
+
+
+def sample_tmsv(
+    sq: list, t: float, Ul: np.ndarray, w: np.ndarray, n_samples: int, loss: float = 0.0,
+) -> list:
+    r"""Generate samples for simulating vibrational quantum dynamics with a two-mode squeezed input state.
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:262]:
```suggestion
    r"""Generate samples for simulating vibrational quantum dynamics with a two-mode squeezed vacuum input state.
```
Note that line width may be too long now

File: strawberryfields/apps/qchem/dynamics.py [hunk 4]
```diff
@@ -250,3 +254,84 @@ def prob(samples: list, excited_state: list) -> float:
...
+
+    Given :math:`N` squeezing parameters and :math:`N`-dim normal-to-local transformation matrix,
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:264]:
```suggestion
    Given :math:`N` squeezing parameters and an :math:`N`-dimensional normal-to-local transformation matrix,
```

File: strawberryfields/apps/qchem/dynamics.py [hunk 5]
```diff
@@ -250,3 +254,84 @@ def prob(samples: list, excited_state: list) -> float:
...
+    a GBS device with :math:`2N` modes is simulated. The evolution operator acts on only the first
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:265]: change to: "this function generates samples from a GBS device with two-mode squeezed vacuum input states."

File: strawberryfields/apps/qchem/dynamics.py [hunk 6]
```diff
@@ -250,3 +254,84 @@ def prob(samples: list, excited_state: list) -> float:
...
+    :math:`N` modes in the device. Samples are generated by measuring the number of photons in each
+    of the :math:`2N` modes.
+
+    **Example usage:**
+
+    >>> sq = [[0.2, 0.1], [0.8, 0.2]]
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:271]:
```suggestion
    >>> r = [[0.2, 0.1], [0.8, 0.2]]
```

File: strawberryfields/apps/qchem/dynamics.py [hunk 7]
```diff
@@ -250,3 +254,84 @@ def prob(samples: list, excited_state: list) -> float:
...
+    >>> t = 10.0
+    >>> Ul = np.array([[0.707106781, -0.707106781],
+    >>>                [0.707106781, 0.707106781]])
+    >>> w = np.array([3914.92, 3787.59])
+    >>> n_samples = 5
+    >>> sample_tmsv(sq, t, Ul, w, n_samples)
+    [[0, 0, 0, 0], [0, 0, 0, 0], [0, 1, 0, 1], [0, 1, 0, 1], [0, 2, 0, 2]]
+
+    Args:
+        sq (list[list[float]]): list of two-mode squeezing gate parameters given as [amplitude, phase] for all modes
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:281]:
```suggestion
        r (list[list[float]]): list of two-mode squeezing gate parameters given as [amplitude, phase] for all modes
```

File: strawberryfields/apps/qchem/dynamics.py [hunk 8]
```diff
@@ -250,3 +254,84 @@ def prob(samples: list, excited_state: list) -> float:
...
+        Ul (array): normal-to-local transformation matrix
+        w (array): normal mode frequencies :math:`\omega` in units of :math:`\mbox{cm}^{-1}`
+        n_samples (int): number of samples to be generated
+        loss (float): loss parameter denoting the fraction of lost photons
+
+    Returns:
+        list[list[int]]: a list of samples
+    """
+    if n_samples < 1:
+        raise ValueError("Number of samples must be at least one")
```
Comment (ixfoduap) [inline strawberryfields/apps/qchem/dynamics.py:292]: Same comment as with the coherent sampling PR. I think some of these checks are unnecessary: time and frequencies are positive by definition and the bounds on the loss parameter should show already from `sf.ops.LossChannel()`

File: strawberryfields/apps/qchem/dynamics.py (hunk 2 above)
Comment (ixfoduap) [reply strawberryfields/apps/qchem/dynamics.py:337]: @trbromley do you agree? I remember we had some issues in the apps layer when using lists instead of arrays

File: strawberryfields/apps/qchem/dynamics.py (hunk 5 above)
Comment (AroosaIjaz) [reply strawberryfields/apps/qchem/dynamics.py:265]: As the beginning of the second sentence in the doc string?

File: strawberryfields/apps/qchem/dynamics.py (hunk 2 above)
Comment (soranjh) [reply strawberryfields/apps/qchem/dynamics.py:333]: We use `return s.tolist()` in `apps.sample`.

File: strawberryfields/apps/qchem/dynamics.py [hunk 9]
```diff
@@ -256,6 +264,80 @@ def prob(samples: list, excited_state: list) -> float:
...
+    >>> t = 10.0
+    >>> Ul = np.array([[0.707106781, -0.707106781],
+    >>>                [0.707106781, 0.707106781]])
+    >>> w = np.array([3914.92, 3787.59])
+    >>> n_samples = 5
+    >>> sample_tmsv(r, t, Ul, w, n_samples)
+    [[0, 0, 0, 0], [0, 0, 0, 0], [0, 1, 0, 1], [0, 1, 0, 1], [0, 2, 0, 2]]
+
+    Args:
+        r (list[list[float]]): list of two-mode squeezing gate parameters given as [amplitude, phase] for all modes
```
Comment (trbromley) [inline strawberryfields/apps/qchem/dynamics.py:291]:
```suggestion
        r (list[list[float]]): list of two-mode squeezing gate parameters given as ``[amplitude, phase]`` for all modes
```

File: strawberryfields/apps/qchem/dynamics.py [hunk 10]
```diff
@@ -256,6 +264,80 @@ def prob(samples: list, excited_state: list) -> float:
...
+            " transformation matrix must be equal"
+        )
+
+    N = len(Ul)
+    op = evolution(N)
+
+    eng = sf.LocalEngine(backend="gaussian")
+    prog = sf.Program(2 * N)
+
+    # pylint: disable=expression-not-assigned,pointless-statement
```
Comment (trbromley) [inline strawberryfields/apps/qchem/dynamics.py:319]: Do we need the `pointless-statement` one?

File: strawberryfields/apps/qchem/dynamics.py (hunk 1 above)
Comment (trbromley) [reply strawberryfields/apps/qchem/dynamics.py:333]: This is also in the standard `sample.sample()` function - the idea is to catch the warning that @AroosaIjaz shared. This warning happens when you have Fock measurements in the Gaussian backend, but I think it's unnecessary in the apps layer.

File: strawberryfields/apps/qchem/dynamics.py (hunk 2 above)
Comment (trbromley) [reply strawberryfields/apps/qchem/dynamics.py:338]: Sorry, only just got around to replying on the other thread! https://github.com/XanaduAI/strawberryfields/pull/423/files#r447988839
TLDR: I think it's ok as it follows what we do in the rest of the apps layer, now including in the `qchem` module, but maybe we can change going forward.

File: tests/apps/qchem/test_dynamics.py [hunk 11]
```diff
@@ -328,6 +234,150 @@ def test_negative_state(self, e):
...
+
+            assert len(dims) == 2
+            assert dims == (tmsv[4], len(tmsv[2]) * 2)
+            assert samples.dtype == "int"
+            assert (samples >= 0).all()
+
+
+@pytest.mark.parametrize("func", sampling_func)
+@pytest.mark.parametrize("par", [d1, tmsv1])
+class CommonTests:
```
Comment (trbromley) [inline tests/apps/qchem/test_dynamics.py:301]: Note that this will have conflicts with https://github.com/XanaduAI/strawberryfields/pull/423 when that is merged in.

**Comment by AroosaIjaz** [comment]
> Thanks @AroosaIjaz, this looks good! Please can we wait until #423 is merged in before doing this one - this will make sure we are on the same page with the tests, as some added content there is repeated here.

yeah, I agree :)

**Comment by AroosaIjaz** [comment]
This PR is merge ready from my side.
